drivers/stepper/TMC2209.zig

```diff
@@ -0,0 +1,576 @@
+            .address = cfg.address,
+            .steps = cfg.steps,
+            .pulse_frequency = cfg.pulse_frequency,
+            .ramp_pause = cfg.ramp_pause,
+        };
+    }
+
+    pub fn spreadcycle(self: *Self) !void {
+        const nc = nodeConf{ .val = .{ .sendDelay = 2 } };
+        try self.write(nc.register, @bitCast(nc.val));
+        const gc = gconf{ .val = .{ .enSpreadcycle = 1, .pdnDisable = 1, .mStepRegSelect = 1, .indexStep = 1 } };
+        try self.write(gc.register, @bitCast(gc.val));
+        const ir = iholdRun{ .val = .{ .ihold = 1, .irun = 31 } };
+        try self.write(ir.register, @bitCast(ir.val));
+        const pw = pwmConf{ .val = .{ .pwmAutoscale = 1, .pwmAutograd = 1, .pwmReg = 8 } };
+        try self.write(pw.register, @bitCast(pw.val));
+    }
+
+    // velocity = hz / 0.715 * steps * microsteps
+    pub fn move(self: *Self, hz: f32) !void {
+        if (hz > 20 or hz < -20) {
+            return error.InvalidMoveFrequency;
+        }
+
+        //uint32((r / 0.715) * float64(m.steps) * float64(m.microsteps))
+        const x: f32 = @floatFromInt((self.steps * self.microsteps));
+        for (self.get_ramp(hz)[0..]) |val| {
+            const v = vactual{
+                .val = .{ .velocity = @intFromFloat(val * (x / self.pulse_frequency)) },
+            };
+
+            try self.write(v.register, @bitCast(v.val));
+            self.clock_device.sleep_ms(self.ramp_pause);
+        }
+        self.current_freq = hz;
+    }
+
+    pub fn set_microsteps(self: *Self, microsteps: u32) !void {
+        self.microsteps = microsteps;
+        const cf = chopconf{
+            .val = .{
+                .toff = 5,
+                .mres = self.mres(microsteps),
+                .intpol = 1,
+            },
+        };
+        try self.write(cf.register, @bitCast(cf.val));
+    }
+
+    pub fn write(self: *Self, register: u8, data: u32) !void {
+        const req = write_request{
+            .addr = self.address,
+            .register = register | 0x80,
+            .data = std.mem.nativeToBig(u32, data),
+        };
+
+        var buf = std.mem.toBytes(req);
+        buf[7] = self.crc(buf[0..7]);
+        const n = try self.uart.write(&buf);
+        if (n != 8) {
+            return error.WriteError;
+        }
+        self.clock_device.sleep_ms(10);
+    }
+
+    pub fn read(self: *Self, register: anytype) !void {
+        const req = read_request{
+            .addr = self.address,
+            .register = register.register & 0x7f,
+        };
+
+        var buf = std.mem.toBytes(req);
+        buf[3] = self.crc(buf[0..3]);
+        const w1 = try self.uart.write(&buf);
+        if (w1 != 4) {
+            return error.InvalidWrite;
+        }
+
+        std.log.debug("read request {x}", .{buf});
+
+        self.clock_device.sleep_ms(100);
+        // The read request will be in the rx buffer since the
+        // tx and rx pins are physically connected.
+        _ = self.uart.read(&buf) catch |err| {
```

Comment:
This read function is giving me an error (error: panic: main() returned error.IoError) and I can't figure out what's wrong.

uart.write works fine (I'm able to move the motor with uart commands).

Reply:
Have you tried with a different mcu?

Reply:
By the way, I have a Go version that works just fine on a rpi-zero for read and write.  It probably doesn't help, but here it is. 

https://github.com/cswank/tmc2209